src/jit/block.h

```diff
@@ -144,6 +144,85 @@ struct EntryState
     StackEntry* esStack;               // ptr to  stack
 };
 
+// Enumeration of the kinds of memory whose state changes the compiler tracks
+enum MemoryKind
+{
+    ByrefExposed = 0, // Includes anything byrefs can read/write (everything in GcHeap, address-taken locals,
+                      //                                          unmanaged heap, callers' locals, etc.)
+    GcHeap,           // Includes actual GC heap, and also static fields
+    MemoryKindCount,  // Number of MemoryKinds
+};
+#ifdef DEBUG
```

Comment:
I think I avoided baking that assumption into the places that use `byrefStatesMatchGcHeapStates`.  I used a few patterns:

1 - places where `output[GcHeap]` and `output[ByrefExposed]` reference the same data structure:
```c++
  for (MemoryKind memoryKind : allMemoryKinds()) {
    if (memoryKind == GcHeap && statesMatch) {
      // already updated
      assert(memoryKind > ByrefExposed);
      assert(input[memoryKind] == input[ByrefExposed]); // and/or assert(output[memoryKind] == f(input[memoryKind]))
      continue;
    }
    output[memoryKind] = f(input[memoryKind]);
  }
```

2a - places where `output[GcHeap]` and `output[ByrefExposed]` need to be copies of each other:
```c++
  for (MemoryKind memoryKind : allMemoryKinds()) {
    if (memoryKind == GcHeap && statesMatch) {
      // copy result from ByrefExposed
      assert(memoryKind > ByrefExposed);
      output[memoryKind] = output[ByrefExposed];
      continue;
    }
    output[memoryKind] = f(input[memoryKind]);
  }
```

2b - other places where `output[GcHeap]` and `output[ByrefExposed]` need to be copies of each other:
```c++
  for (MemoryKind memoryKind : allMemoryKinds()) {
    if (memoryKind == GcHeap && statesMatch) {
      // already updated
      assert(memoryKind > ByrefExposed);
      assert(input[memoryKind] == input[ByrefExposed]); // and/or assert(output[memoryKind] == f(input[memoryKind]))
      continue;
    }
    output[memoryKind] = f(input[memoryKind]);
    if (memoryKind == GcHeap && statesMatch) {
      // copy over to ByrefExposed
      output[ByrefExposed] = output[memoryKind];
    }
  }
```

(2a was more convenient for some of the particular updates, 2b for others).

In all cases, if a third kind were added the loop would just process it independently of GcHeap and ByrefExposed, which I think is a good default position.

Does that address your concern, or am I misunderstanding the feedback?